### PR TITLE
feat: add esbuild plugin in webpack-builder

### PR DESCRIPTION
**Pull request by sanyuan0704**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/plugin-esbuild/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import type { BuilderPlugin } from '@modern-js/webpack-builder/dist/types';
```
Comment (chenjiahan) [inline packages/builder/plugin-esbuild/src/index.ts:1]: `@modern-js/webpack-builder/dist/types` -> `@modern-js/webpack-builder`

File: packages/builder/plugin-esbuild/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  minimize?: MinifyPluginOptions;
+}
+
+export const JS_REGEX = /\.(js|mjs|cjs|jsx)$/;
+export const TS_REGEX = /\.(ts|mts|cts|tsx)$/;
+
+export function PluginEsbuild(
+  options: EsbuildOptions = {
+    loader: {},
+    minimize: {},
```
Comment (chenjiahan) [inline packages/builder/plugin-esbuild/src/index.ts:18]: loader 和 minimize 有开关不，对于使用者来说，应该可以选择使用 esbuild minify 或 esbuild transform 中的某一项

File: packages/builder/webpack-builder/tsconfig.json [hunk 3]
```diff
@@ -9,5 +9,5 @@
     "isolatedModules": true,
     "paths": {}
   },
-  "include": ["src"]
+  "include": ["src", "./shims.d.ts"]
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/tsconfig.json:12]: shims.d.ts 好像没看到

File: packages/toolkit/utils/src/chainId.ts [hunk 4]
```diff
@@ -130,6 +132,8 @@ export const CHAIN_ID = {
     NODE_POLYFILL_PROVIDE: 'node-polyfill-provide',
     /** ModuleDependencyErrorPlugin */
     MODULE_DEPENDENCY_ERROR: 'module-dependency-error',
+    /** EsbuildMinifyPlugin */
+    ESBUILD_MINIFY: 'esbuild-minify',
```
Comment (chenjiahan) [inline packages/toolkit/utils/src/chainId.ts:136]: 这个常量似乎没用到

File: packages/builder/webpack-builder/tsconfig.json (hunk 3 above)
Comment (sanyuan0704) [reply packages/builder/webpack-builder/tsconfig.json:12]: 这里我清理下吧

File: packages/builder/plugin-esbuild/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              loader: 'tsx',
+              target: 'es2015',
+              ...options?.loader,
+            });
+        }
+
+        if (isProd && options.minimize !== false) {
+          const { ESBuildMinifyPlugin } = (await import(
+            compiledEsbuildLoaderPath
+          )) as typeof import('../compiled/esbuild-loader');
```
Comment (chenjiahan) [inline packages/builder/plugin-esbuild/src/index.ts:62]: 这里 as typeof import 有点冗余了，不如直接 import

```ts
await import('/../compiled/esbuild-loader');
```

File: packages/builder/plugin-esbuild/src/index.ts (hunk 2 above)
Comment (sanyuan0704) [reply packages/builder/plugin-esbuild/src/index.ts:18]: 可以通过配置 loader: false  或者 minimize: false 来关闭

File: packages/builder/plugin-esbuild/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          chain.optimization
+            .minimizer(CHAIN_ID.PLUGIN.ESBUILD_MINIFY)
```
Comment (chenjiahan) [inline packages/builder/plugin-esbuild/src/index.ts:64]: 需要把内置的 minimizer 删除

```ts
chain.optimization.minimizers
  .delete(CHAIN_ID.MINIMIZER.JS)
  .delete(CHAIN_ID.MINIMIZER.CSS);
```

File: packages/builder/plugin-esbuild/src/index.ts (hunk 5 above)
Comment (sanyuan0704) [reply packages/builder/plugin-esbuild/src/index.ts:62]: 确实，我也纠结了一阵

File: packages/builder/plugin-esbuild/src/index.ts (hunk 6 above)
Comment (sanyuan0704) [reply packages/builder/plugin-esbuild/src/index.ts]: 是的，应该用 chain.optimization.delete

**Review by chenjiahan — APPROVED** [review]

**Comment by sanyuan0704** [comment]
🎉

**Pull request merged**